pkg/mimirpb/split.go

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package mimirpb
+
+// SplitWriteRequestByMaxMarshalSize splits the WriteRequest into multiple ones, where each partial WriteRequest marshalled size
+// is at most maxSize. The input reqSize must be the value returned by WriteRequest.Size(); it's passed because the called
+// may have already computed it.
+//
+// This function guarantees that a single Timeseries or Metadata entry is never split across multiple requests.
+// For this reason, this function is a best-effort: if a single Timeseries or Metadata marshalled size is bigger
+// than maxSize, then the returned partial WriteRequest marshalled size will be bigger than maxSize too.
+//
+// The returned partial WriteRequests are NOT a deep copy of the input one; they contain references to slices
+// and data from the original WriteRequest.
+func SplitWriteRequestByMaxMarshalSize(req *WriteRequest, reqSize, maxSize int) []*WriteRequest {
+	if reqSize <= maxSize {
+		return []*WriteRequest{req}
+	}
+
+	partialReqsWithTimeseries := splitTimeseriesByMaxMarshalSize(req, reqSize, maxSize)
+	partialReqsWithMetadata := splitMetadataByMaxMarshalSize(req, reqSize, maxSize)
+
+	// Most of the time, a write request only have either Timeseries or Metadata.
+	if len(partialReqsWithMetadata) == 0 {
+		return partialReqsWithTimeseries
+	}
+	if len(partialReqsWithTimeseries) == 0 {
+		return partialReqsWithMetadata
+	}
+
+	merged := make([]*WriteRequest, 0, len(partialReqsWithTimeseries)+len(partialReqsWithMetadata))
+	merged = append(merged, partialReqsWithTimeseries...)
+	merged = append(merged, partialReqsWithMetadata...)
+
+	return merged
+}
+
+func splitTimeseriesByMaxMarshalSize(req *WriteRequest, reqSize, maxSize int) []*WriteRequest {
+	if len(req.Timeseries) == 0 {
+		return nil
+	}
+
+	newPartialReq := func(preallocTimeseries int) (*WriteRequest, int) {
+		r := &WriteRequest{
+			Timeseries:              preallocSliceIfNeeded[PreallocTimeseries](preallocTimeseries),
```

Comment:
We don't need to allocate `Timeseries` slice, if we never write to that slice. We *always* **replace** it with slices from original request.

(Same comment applies in `splitMetadataByMaxMarshalSize`).

Reply:
Right! I changed the logic while working on the benchmark, and then forgot to remove the preallocation.

Reply:
Fixed in https://github.com/grafana/mimir/pull/8077/commits/5126f036773a3ab37ecad5ee5d3654ffd061bdb5